serverless/api/state.go

```diff
@@ -24,6 +35,44 @@ type LogState struct {
 	RootHash []byte
 }
 
+// MarshalText implements golang's encoding/TextMarshaler interface, and renders
+// the LogState into a UTF8 text based format:
+//
+// Log Checkpoint v0\n
+// <decimal log size>\n
+// <base64 encoded root hash>\n
+func (s LogState) MarshalText() ([]byte, error) {
+	r := fmt.Sprintf("%s\n%d\n%s\n", CheckpointHeader, s.Size, base64.StdEncoding.EncodeToString(s.RootHash))
+	return []byte(r), nil
+}
+
+// UnmarshalText implements golang's encoding/TextUnmarshaler interface, and
+// populates this struct with the parsed values from the supplied text.
+//
+// This method reads the data serialised by the MarshalText method above,
+// and it ignores any trailing.
+func (s *LogState) UnmarshalText(raw []byte) error {
+	text := string(raw)
+	l := strings.Split(text, "\n")
+	if len(l) < 3 {
+		return errors.New("invalid checkpoint - too few lines")
+	}
+	if l[0] != CheckpointHeader {
+		return errors.New("invalid checkpoint - incorrect header")
+	}
+	size, err := strconv.ParseUint(l[1], 10, 64)
+	if err != nil {
+		return fmt.Errorf("invalid checkpoint - size invalid: %w", err)
+	}
+	rh, err := base64.StdEncoding.DecodeString(l[2])
+	if err != nil {
+		return fmt.Errorf("invalid checkpoint - invalid roothash: %w", err)
+
+	}
+	s.Size, s.RootHash = size, rh
```

Comment:
Consider checking the size of `rh`.  The doc for `RootHash` says it's 256 bits so may as well be paranoid on inputs. Though I realize this totally makes your tests less fun. So YMMV :-)

Reply:
I think that's probably the layer above (unless I make the struct have a `[32]byte` for `RootHash`?), maybe I'll update the comment instead :)